Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -274,83 +273,101 @@ smoothLogSieveM sievingIntervalM factorBaseWithSquareRoots a b c m =
-        -- the factorisation with respect to the primes in the @factorBase@ and it appends
-        -- the remainder of the number after all the divisions ifthis is larger than one.
-        -- Note that this is in ascending order since the remainder cannot be smaller than @highestPrime@.
-        listFactorisation = trialDivisionWith (map (intToInteger . unPrime) factorBase) value
+        -- This returns the highest factor in a factorisation whenever this
+        -- is of Int type and prime. Otherwise returns @Nothing@.
+        -- The maximum in @preFac@ is the number that is left after dividing by
+        -- all the primes in @factorBase@. Note that @preFac@ is empty whenever
+        -- @value = 1@, hence the need to prepend @1@.
+        highestFactor = isPrime =<< (toIntegralSized (maximum (1 : fmap fst preFac)) :: Maybe Int)
+        -- The conversion to Int is safe since @facMap@ is only used whenever
+        -- @highestFactor@ is itself an Int.
+        facMap = IM.fromDistinctAscList $ map (bimap integerToInt wordToInt) fullFac
+        -- Add negative unit.
+        fullFac = if value < 0 then (-1, 1) : preFac else preFac
+        -- This performs trial division with prime numbers in the @factorBase@.
+        -- It returns the factorisation with respect to the primes in the
+        -- @factorBase@ and it appends the remainder of the number after all the
+        -- divisions, if this is larger than one. Note that the resulting list
+        -- is necessarily in ascending order since the remainder cannot be
+        -- smaller than @highestPrime@.
+        preFac = trialDivisionWith (map (intToInteger . unPrime) factorBase) value
 
+    -- This function completes the square in @f@.
+    complete i = a * intToInteger (i - m) + b
     a = factorBack decompositionOfA
-    intMapA = I.fromAscList $ map (bimap (integerToInt . unPrime) wordToInt) decompositionOfA
-    highestPrime = intToInteger . unPrime . maximum $ factorBase
+    intMapA = IM.fromDistinctAscList $ map (bimap (integerToInt . unPrime) wordToInt) decompositionOfA
 
--- Removes all columns of the matrix which contain primes appearing only once.
--- These columns cannot be part of the solution.
-removeRows :: [(Integer, I.IntMap Int)] -> [(Integer, I.IntMap Int)]
-removeRows indexedFactorisations
-  | onlyOnce == mempty = indexedFactorisations
-  | otherwise          = removeRows $ filter (\(_, im) -> S.null (S.intersection (convertToSet im) onlyOnce)) indexedFactorisations
+-- Removes all columns of the matrix which contain primes appearing only once
+-- and are greater than @lowerBound@. When @lowerBound = 0@, this routine
+-- removes columns which cannot be part of the solution.
+removeColumns :: Int -> M.Map Integer (IM.IntMap Int) -> M.Map Integer (IM.IntMap Int)
+removeColumns lowerBound smoothData
+  | onlyOnce == mempty = smoothData
+  | otherwise          = removeColumns lowerBound $ M.filter (IS.null . IS.intersection onlyOnce . setOddPowers) smoothData
```

Comment:
Let's define `disjoint` equal to `IS.disjoint` for newer `containers` and to `IS.null . IS.intersection` for older.